Only change expiration date if it is different

samdoran opened the pull request:
##### SUMMARY

Modify `user_info()` method to also return the password expiration.
Compare current and desired expiration times and only change if they are different.

Fixes #13235

<!--- Describe the change, including rationale and design decisions -->

<!--- If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request


##### COMPONENT NAME
`user.py`

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
2.5
```

samdoran committed:
Only change expiration date if it is different

Modify user_info() method to also return the password expiration.
Compare current and desired expiration times and only change if they are different.

samdoran committed:
Improve formatting on user tests

samdoran committed:
Add integration test for expiration

samdoran committed:
Add changelog fragment

webknjaz commented:
-label needs_triage

bcoca commented:
@sdoran, i believe the failing tests are wrong

samdoran commented:
@bcoca Looks like some tests are incorrect, others are legit.

I need to skip for macOS. The [failure](https://app.shippable.com/github/ansible/ansible/runs/61682/28/tests) on CentOS 6 with Python 2.6 is legit: no `totalseconds()` method ☹️. Guess I'll have to do that differently.

Also, I need to add expiration change logic for FreeBSD as well. I missed that since I only fixed the method in the User class.

Some of the tests, I'm not sure what is triggering the change on the second task run. [RHEL7.4/1](https://app.shippable.com/github/ansible/ansible/runs/61682/25/tests) for example.

samdoran committed:
Improve integration test

Skip macOS and use getent module for validating expiration date.

samdoran committed:
Fix expiration change for FreeBSD

samdoran committed:
Don't use datetime since the total_seconds method isn't available on CentOS 6

bcoca commented on lib/ansible/modules/system/user.py:
```diff
@@ -271,6 +270,7 @@ class User(object):
     platform = 'Generic'
     distribution = None
     SHADOWFILE = '/etc/shadow'
+    SHADOWFILE_INDEX = 7
```
shadowfile_expire_index? .. just index seems ambiguous and we might need others in future

samdoran replied:
Very good point.

samdoran committed:
Use better name for expiration index field

Use separate tasks for verifying expiration date on BSD

bcoca commented on test/integration/targets/user/tasks/main.yml:
```diff
@@ -20,142 +20,200 @@
-
-##
+    that:
+      - "not user_test1.results[0]['changed']"
```
the proper test would be ` is changed`

samdoran commented:
@bcoca So the failing tests revealed a couple problems:

1. I used the incorrect date math functions that mix localtime and UTC.
2. The `timezone` integration tests change the timezone on the system and don't set it back to UTC.

So the untidy `timezone` test actually exposed a bug in my fix. Lucky for me.

I'm working on getting this fixed now that I've been able to figure out what's going on (with a ton of help from @mattclay).

samdoran committed:
Use calendar.timegm() rather than time.mktime()

calendar.timegm() is the inverse of time.gmtime() and returns a timestamp in UTC not localtime
Add tests that change the system timezone away from UTC

samdoran committed:
Mark tests as destructive and use test for change status

mattclay commented:
CI failure in FreeBSD integration tests: 

https://app.shippable.com/github/ansible/ansible/runs/62723/26/tests
https://app.shippable.com/github/ansible/ansible/runs/62723/27/tests

The other failures are unrelated to this PR.

samdoran committed:
Fix account expiration for FreeBSD

Use DATE_FORMAT when setting expiration date on FreeBSD. Previously the argument passed to -e was an integer of days since epoch when the account will expire which was inserted directly into master.passwd. This value is interpreted as seconds since epoch by the system, meaning the account expiration was actually set to a few hours past epoch.

Greatly simply comparing desired  and current expiration time by using the first three values of the struct_time tuple rather than doing a whole bunch of manipulations of the seconds since epoch.

samdoran commented:
@bcoca I think I finally got this working.

I found a bug in the FreeBSD class. Passing an int to `-e` sets that directly in `/etc/master.passwd`:

```
root@ip-192-168-3-46:~/ansible$ grep ansi /etc/master.passwd
ansibulluser:*:1002:1002::0:6221:User &:/home/ansibulluser:/bin/sh
                            ^
                            This is seconds since epoch
```

The man pages aren't clear and some of them do say an integer in DAYS or a date string. I switched to using the date string since that seems to behave correctly.
